Map/LightOverlay.gd

```diff
@@ -0,0 +1,77 @@
+extends Node2D
+class_name LightOverlay
+
+var color_array: PackedColorArray
+
+var room_container: ColorRect
+var room_ui_array: Array[Array]
+
+var room_polygons_to_draw: Array[Rect2]
+
+func _init(_room_container: ColorRect, _room_ui_array: Array[Array]):
+	room_container = _room_container
+	room_ui_array = _room_ui_array
+
+func _draw():
+	var floor_rect_packed_array: PackedVector2Array
+	var room_circles: Array[PackedVector2Array]
+	var lit_rooms_to_draw: Array[RoomUI]
+	
+	# Account for minimum map placement T____T
+	for floor_index: int in range(room_ui_array.size()):
+		var floor_array: Array = room_ui_array[floor_index]
+		var _first_room_in_array: bool = false
+		
+		for room_index: int in range(floor_array.size()):
+			var room: RoomUI = floor_array[room_index]
+			if room != null:
+				var new_room_texture_rect: TextureRect = Helpers.get_first_child_node_of_type(room, TextureRect)
+				var room_size = new_room_texture_rect.get_size()
+				if (room.get_light_level() != Enums.LightLevel.UNLIT):
+					room_circles.append(_calculate_points_for_circle(room))
+					lit_rooms_to_draw.append(room)
+	floor_rect_packed_array.append(Vector2(0, 0))
+	floor_rect_packed_array.append(Vector2(0, room_container.get_size().y))
+	floor_rect_packed_array.append(Vector2(room_container.get_size().x, room_container.get_size().y))
+	floor_rect_packed_array.append(Vector2(room_container.get_size().x, 0))
```

Comment:
what is the meaning of those 4 vectors being added ?

Reply:
I've left a comment in the file, but this'll also be updated as well to be split up.

These vectors are used to help find the polygon of the darkness overlay that surrounds the lit rooms.

I've updated the logic a bit from this comment, but I'll explain it in detail here:
We have two different packed arrays one for the upper half of the map at the player position (In red), and one for the bottom half of the map at the player position (In green). Anything that is in black in this example picture is the lit room polygon that is calculated in Line 60-64

Geometry2D.clip_polygons() is a function in Godot that will return the inverse polygon of two polygons that are passed in. HOWEVER, this only works if there is OVERLAP, if the polygons are fully encompassed or this is no overlap, then we won't get any polygons returned.
![image](https://github.com/Saplings-Projects/1M_sub/assets/5923102/b21918af-1cca-495c-8155-26061f74fce7)

So we get the inverse of the top half, and draw that out (Line 67-69):
![image](https://github.com/Saplings-Projects/1M_sub/assets/5923102/76493d46-58ce-4393-8990-820cdf8f55f9)

Then we get inverse of the bottom half, and draw that out (Line 70-72):
![image](https://github.com/Saplings-Projects/1M_sub/assets/5923102/b4386448-f19c-430e-9379-9f5fe79f8eec)

Reply:
I realized the naming's confusing, so I'll update it alongside some comments.

Reply:
So basically you cut the map in half to draw a circle around the player ?

Reply:
To draw a circle around all of the lit rooms.

Reply:
Oh so you move the place where it's cut in half when you go to the next floor ?

Reply:
Not exactly. We just get the very first room. We just need need to cut it in half anywhere there's a circle in order to be able to get the inverse shape, since Geometry2D.clip_polygons will only return something if there's an intersection and not if it's fully covered.